pkg/endpoint/endpoint.go

```diff
@@ -361,14 +397,122 @@ func (e *Endpoint) WaitForProxyCompletions() error {
 	return nil
 }
 
+// runK8sCiliumEndpointSync starts a controller that syncronises the endpoint
+// to the corresponding k8s CiliumEndpoint CRD
+// ep.Mutex must be help for the setup portion of this function, but not the
+// controller.
+// CiliumEndpoint objects follow a nodename-cep-endpointID scheme
+func (e *Endpoint) runK8sCiliumEndpointSync() {
+	var (
+		endpointID     = e.ID
+		controllerName = fmt.Sprintf("sync-to-k8s-ciliumendpoint (%v)", endpointID)
+		epName         = fmt.Sprintf("%v-cep-%v", node.GetName(), endpointID)
+		scopedLog      = e.getLogger().WithField("controller", controllerName)
+	)
+
+	if !k8s.IsEnabled() {
+		scopedLog.Debug("Not starting controller because k8s is disabled")
+		return
+	}
+
+	// The health endpoint is created outside of k8s and shouldn't create a CEP
+	// for it. It has a specific label on it and we do not start the controller
+	// in that case.
+	healthLbl := labels.NewLabel(labels.IDNameHealth, "", labels.LabelSourceReserved)
+	for _, lbl := range e.OpLabels.OrchestrationIdentity {
+		if healthLbl.Equals(lbl) {
+			scopedLog.Debug("Not starting controller because it is the reserved internal health endpoint")
+			return
+		}
+	}
+
+	ciliumClient, err := getCiliumClient()
+	if err != nil {
+		scopedLog.WithError(err).Error("Not starting controller because unable to get cilium k8s client")
+		return
+	}
+
+	var lastMdl *models.Endpoint
+
+	// NOTE: The controller functions do NOT hold the endpoint locks
+	e.controllers.UpdateController(controllerName,
+		controller.ControllerParams{
+			RunInterval: 30 * time.Second,
+			DoFunc: func() error {
+				namespace := e.GetK8sNamespace()
+				if namespace == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has no k8s namespace")
+					return nil
+				}
+
+				mdl := e.GetModel()
+				if reflect.DeepEqual(mdl, lastMdl) {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has not changed")
+					return nil
+				}
+				lastMdl = mdl
```

Comment:
It can still fail in which case we will never update again, I would update this last

Reply:
The sad part is that this check will never actually do anything. We include the controller status in mdl but this is a controller, so every iteration we see a different value from the last time. I don't know how to fix this without suppressing the controller information. That would undermine the point (to some extent). Another would be to ignore the status for this controller but it would depend on the frequency of other controllers then. What do you think?

Reply:
You could store the last model in the endpoint

Reply:
Could you explain more what you're thinking? Obviously we can store the data but I'm not sure why it would be useful overall. Since we didn't have a `data` parameter on the DoFunc I assumed we intended to use the fact that you keep the state within the closure you create for DoFunc. As long as we're consistent, I like the idea of keeping controller state close to the functions themselves since it makes them far more self-contained (The only rebase conflicts I've had this PR have been imports, not the actual important code). This falls apart if they start editing fields on the endpoint willy nilly, however.